### PR TITLE
CPLP-1212 services

### PR DESCRIPTION
**Pull request by Phil91**
In this PR the Tasks [CPLP-1214](https://jira.catena-x.net/browse/CPLP-1214) [CPLP-1215](https://jira.catena-x.net/browse/CPLP-1215) [CPLP-1216](https://jira.catena-x.net/browse/CPLP-1216) [CPLP-1213](https://jira.catena-x.net/browse/CPLP-1213) [CPLP-1413](https://jira.catena-x.net/browse/CPLP-1413) & [CPLP-1427](https://jira.catena-x.net/browse/CPLP-1427) are resolved.

Please review the database changes in [PR](https://github.com/tractusx-team-portal-onboarding/devenv/pull/114) as well

### CODE REVIEW

**Comment by Phil91** [comment]
@ntruchsess in my opinion it will make sense to finish the #183 first. The Vulnerability, Security Hotspot and duplication will disappear than.

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/AppRepository.cs [hunk 1]
```diff
@@ -215,4 +215,39 @@ public  Task<(IEnumerable<AppDescription> descriptions, IEnumerable<AppDetailIma
...
                        ))
              .SingleOrDefaultAsync();
+
+     /// <inheritdoc />
+     public IQueryable<ServiceDetailData> GetActiveServices() =>
+         _context.Apps
+             .Where(x => x.AppTypeId == AppTypeId.SERVICE && x.AppStatusId == AppStatusId.ACTIVE)
+             .Select(app => new ServiceDetailData(
+                 app.Id,
+                 app.Name ?? Constants.ErrorString,
```
Comment (ntruchsess) [inline src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/AppRepository.cs:225]: please to the mapping to ErrorString in the businesslogic. Doing it on the database-server just enlarges to payload with no benefit. (And take care of propper formating: indentation is 4 chars)

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/ILanguageRepository.cs [hunk 2]
```diff
@@ -8,4 +8,11 @@ public interface ILanguageRepository
     /// <param name="languageShortName">the short name of the language</param>
     /// <returns>the shortname if existing otherwise false</returns>
     Task<string?> GetLanguageAsync(string languageShortName);
+
+    /// <summary>
+    /// Checks whether the given language codes exists in the persistence storage
+    /// </summary>
+    /// <param name="languageCodes">the language codes that should be checked</param>
+    /// <returns>Returns the found language codes</returns>
+    Task<List<string>> GetLanguageCodesUntrackedAsync(ICollection<string> languageCodes);
```
Comment (ntruchsess) [inline src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/ILanguageRepository.cs:17]: in the repository we return IAsyncEnumerable, any materialization to list shall be done in businesslogic.

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/UserRepository.cs [hunk 3]
```diff
@@ -324,6 +324,12 @@ public UserRepository(PortalDbContext portalDbContext)
             .Select(cu => new ValueTuple<Guid, string>(cu.Id, cu.Email!))
             .SingleAsync();
 
+    /// <inheritdoc />
+    public IAsyncEnumerable<(Guid CompanyUserId, bool IsIamUser, string CompanyShortName, Guid CompanyId)> GetCompanyUserWithIamUserCheckAndCompanyShortName(string iamUserId, Guid salesManagerId) => 
+            _dbContext.CompanyUsers.Where(x => x.IamUser!.UserEntityId == iamUserId || x.Id == salesManagerId)
```
Comment (ntruchsess) [inline src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/UserRepository.cs:329]: please do propper formating. indentation is 4 chars, not 8.

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.Migrations/Migrations/20220816061525_CPLP-1213-AddServices.cs [hunk 4]
```diff
@@ -279,12 +279,12 @@ protected override void Up(MigrationBuilder migrationBuilder)
                 table: "services",
                 column: "service_status_id");
 
-            migrationBuilder.AddAuditTrigger<AuditService>("audit_services_cplp_1213_add_services");
+            // migrationBuilder.AddAuditTrigger<AuditService>("audit_services_cplp_1213_add_services");
```
Comment (ntruchsess) [inline src/portalbackend/CatenaX.NetworkServices.PortalBackend.Migrations/Migrations/20220816061525_CPLP-1213-AddServices.cs:282]: no commented code in the git-repo, please delete

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.PortalEntities/Entities/AppType.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
+﻿using System.ComponentModel.DataAnnotations;
+using CatenaX.NetworkServices.PortalBackend.PortalEntities.Enums;
+
+namespace CatenaX.NetworkServices.PortalBackend.PortalEntities.Entities;
+
+public class AppType
```
Comment (ntruchsess) [inline src/portalbackend/CatenaX.NetworkServices.PortalBackend.PortalEntities/Entities/AppType.cs:6]: We should consider a rename of the entity-base-name 'App' to something more generic, like 'offer'. Both App and Services do have some properties in common which belong to the base-entiy. Some properties will reside in type-spefic tables (for apps and services) that are just linked to the base entity.
AppType would then be renamed to 'OfferType'.
Maybe you have a better suggestion than 'offer'?

File: src/services/CatenaX.NetworkServices.Service.Service/BusinessLogic/ServiceSettings.cs [hunk 6]
```diff
@@ -18,22 +18,25 @@
...
-    public AuditOperationId AuditOperationId { get; set; }
+public static class ServiceSettingsExtension
+{
+    public static IServiceCollection ConfigureServiceSettings(
+        this IServiceCollection services,
+        IConfigurationSection section)
+    {
+        services.AddOptions<ServiceSettings>()
+            .Bind(section)
+            .ValidateOnStart();
```
Comment (ntruchsess) [inline src/services/CatenaX.NetworkServices.Service.Service/BusinessLogic/ServiceSettings.cs:39]: does ValidateOnStart() do anything if there is no metadata-validation triggered and no custom validation attached?

**Review by ntruchsess — CHANGES_REQUESTED** [review]

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.PortalEntities/Entities/AppType.cs (hunk 5 above)
Comment (Phil91) [reply src/portalbackend/CatenaX.NetworkServices.PortalBackend.PortalEntities/Entities/AppType.cs:6]: Offer was the first thing which came to my mind as well. an alternative would be component but that's not really matching in my opinion. I've talked to Julia before she went on vacation. we wanted to postpone the renaming of the app. maybe we can have a short call about that.

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.Migrations/Migrations/20220816061525_CPLP-1213-AddServices.cs (hunk 4 above)
Comment (Phil91) [reply src/portalbackend/CatenaX.NetworkServices.PortalBackend.Migrations/Migrations/20220816061525_CPLP-1213-AddServices.cs]: removed

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/UserRepository.cs (hunk 3 above)
Comment (Phil91) [reply src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/UserRepository.cs]: done

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/ILanguageRepository.cs (hunk 2 above)
Comment (Phil91) [reply src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/ILanguageRepository.cs]: done

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/AppRepository.cs (hunk 1 above)
Comment (Phil91) [reply src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/AppRepository.cs]: done

File: src/services/CatenaX.NetworkServices.Service.Service/BusinessLogic/ServiceSettings.cs (hunk 6 above)
Comment (Phil91) [reply src/services/CatenaX.NetworkServices.Service.Service/BusinessLogic/ServiceSettings.cs:43]: Tested it with removing the ApplicationsMaxPageSize from the appsettings (it's the only place where it is defined), no error was thrown. I've added the ValidateDataAnnotations

File: src/services/CatenaX.NetworkServices.Service.Service/BusinessLogic/IServiceBusinessLogic.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    /// <returns>The id of the newly created service</returns>
+    Task<Guid> CreateServiceOffering(ServiceOfferingData data, string iamUserId);
+
+    /// <summary>
+    /// Adds a subscription to the given service
+    /// </summary>
+    /// <param name="serviceId">Id of the service the users company should be subscribed to</param>
+    /// <param name="iamUserId">Id of the user</param>
+    /// <returns></returns>
+    Task AddServiceSubscription(Guid serviceId, string iamUserId);
```
Comment (ntruchsess) [inline src/services/CatenaX.NetworkServices.Service.Service/BusinessLogic/IServiceBusinessLogic.cs:51]: shouldn't this return the id of the newly created ordersubscription so the endpoint might return a CreatedAt-response pointing to the new subscription?

File: src/services/CatenaX.NetworkServices.Service.Service/CatenaX.NetworkServices.Service.Service.csproj [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        <ImplicitUsings>enable</ImplicitUsings>
+        <UserSecretsId>9e7888a2-da8b-44ec-8f8b-bd9b4d374726</UserSecretsId>
+        <DockerDefaultTargetOS>Linux</DockerDefaultTargetOS>
+        <DockerfileContext>..\..\..</DockerfileContext>
+        <GenerateDocumentationFile>True</GenerateDocumentationFile>
+        <NoWarn>CS1591</NoWarn>
+    </PropertyGroup>
+
+    <ItemGroup>
+      <ProjectReference Include="..\..\framework\CatenaX.NetworkServices.Framework.Web\CatenaX.NetworkServices.Framework.Web.csproj" />
```
Comment (ntruchsess) [inline src/services/CatenaX.NetworkServices.Service.Service/CatenaX.NetworkServices.Service.Service.csproj:15]: please remove the redundant dependencies that FrameWork.Web implicitly depends on.

**Review by ntruchsess — CHANGES_REQUESTED** [review]

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/OfferRepository.cs [hunk 9]
```diff
@@ -83,9 +83,9 @@ public App CreateApp(string provider, Action<App>? setOptionalParameters = null)
                 ShortDescription =
                     _context.Languages.SingleOrDefault(l => l.ShortName == languageShortName) == null 
                         ? null 
-                        : a.AppDescriptions.SingleOrDefault(d => d.LanguageShortName == languageShortName)!.DescriptionShort
-                          ?? a.AppDescriptions.SingleOrDefault(d => d.LanguageShortName == Constants.DefaultLanguage)!.DescriptionShort,
-                LicenseText = a.AppLicenses
+                        : a.OfferDescriptions.SingleOrDefault(d => d.LanguageShortName == languageShortName)!.DescriptionShort
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/OfferRepository.cs]: fixed

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/LanguageRepository.cs [hunk 10]
```diff
@@ -26,4 +26,11 @@ public LanguageRepository(PortalDbContext portalDbContext)
             .Where(language => language.ShortName == languageShortName)
             .Select(language => language.ShortName)
             .SingleOrDefaultAsync();
+
+    /// <inheritdoc />
+    public IAsyncEnumerable<string> GetLanguageCodesUntrackedAsync(ICollection<string> languageCodes) =>
+        _portalDbContext.Languages.AsNoTracking()
+            .Where(x => languageCodes.Any(y => y == x.ShortName))
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/LanguageRepository.cs]: fixed

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/OfferRepository.cs [hunk 11]
```diff
@@ -119,15 +119,15 @@ public async Task<AppDetailsData> GetAppDetailsByIdAsync(Guid appId, string iamU
                 LongDescription =
                     _context.Languages.SingleOrDefault(l => l.ShortName == languageShortName) == null
                     ? null
-                    : a.AppDescriptions.SingleOrDefault(d => d.LanguageShortName == languageShortName)!.DescriptionLong
-                      ?? a.AppDescriptions.SingleOrDefault(d => d.LanguageShortName == Constants.DefaultLanguage)!.DescriptionLong,
-                Price = a.AppLicenses
+                    : a.OfferDescriptions.SingleOrDefault(d => d.LanguageShortName == languageShortName)!.DescriptionLong
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/OfferRepository.cs]: fixed

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.Migrations/Migrations/20220905140741_CPLP-1212-ChangeAppToOffer.cs [hunk 12]
```diff
@@ -0,0 +1,1417 @@
...
+                    table.PrimaryKey("pk_app_subscription_details", x => x.id);
+                    table.ForeignKey(
+                        name: "fk_app_subscription_details_app_instances_app_instance_id",
+                        column: x => x.app_instance_id,
+                        principalSchema: "portal",
+                        principalTable: "app_instances",
+                        principalColumn: "id");
+                });
+
+            migrationBuilder.CreateTable(
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/portalbackend/CatenaX.NetworkServices.PortalBackend.Migrations/Migrations/20220905140741_CPLP-1212-ChangeAppToOffer.cs]: removed

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.PortalEntities/AuditEntities/AuditOfferSubscription.cs [hunk 13]
```diff
@@ -27,7 +27,7 @@ namespace CatenaX.NetworkServices.PortalBackend.PortalEntities.AuditEntities;
 /// <summary>
 /// Audit entity for App subscription relationship between companies and apps.
 /// </summary>
-public class AuditCompanyAssignedApp : CompanyAssignedApp, IAuditEntity
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/portalbackend/CatenaX.NetworkServices.PortalBackend.PortalEntities/AuditEntities/AuditOfferSubscription.cs:30]: i removed the usings of the Audit entity, but didn't delete it, so we don't forget we need to implement the auditing for it

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.PortalEntities/Entities/OfferSubscription.cs [hunk 14]
```diff
@@ -85,27 +92,24 @@ public CompanyAssignedApp(Guid id, Guid appId, Guid companyId, AppSubscriptionSt
     
     /// <inheritdoc />
     public Guid? LastEditorId { get; set; }
-    
-    [MaxLength(255)]
-    public string? AppUrl { get; set; }
+
+    public Guid? AppSubscriptionDetailId { get; set; }
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/portalbackend/CatenaX.NetworkServices.PortalBackend.PortalEntities/Entities/OfferSubscription.cs]: done

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.PortalEntities/Entities/OfferDetailImage.cs [hunk 15]
```diff
@@ -2,26 +2,26 @@
...
+public class OfferDetailImage
 {
-    private AppDetailImage()
+    private OfferDetailImage()
     {
         ImageUrl = null!;
     }
 
-    public AppDetailImage(Guid appId, string imageUrl)
+    public OfferDetailImage(Guid offerId, string imageUrl)
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/portalbackend/CatenaX.NetworkServices.PortalBackend.PortalEntities/Entities/OfferDetailImage.cs]: done

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.PortalEntities/PortalDbContext.cs [hunk 16]
```diff
@@ -41,21 +41,14 @@ public PortalDbContext(DbContextOptions<PortalDbContext> options)
     public virtual DbSet<AgreementAssignedCompanyRole> AgreementAssignedCompanyRoles { get; set; } = default!;
     public virtual DbSet<AgreementAssignedDocumentTemplate> AgreementAssignedDocumentTemplates { get; set; } = default!;
     public virtual DbSet<AgreementCategory> AgreementCategories { get; set; } = default!;
-    public virtual DbSet<App> Apps { get; set; } = default!;
     public virtual DbSet<AppInstance> AppInstances { get; set; } = default!;
-    public virtual DbSet<AppAssignedLicense> AppAssignedLicenses { get; set; } = default!;
+    public virtual DbSet<AppAssignedDocument> AppAssignedDocuments { get; set; } = default!;
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/portalbackend/CatenaX.NetworkServices.PortalBackend.PortalEntities/PortalDbContext.cs]: renamed

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/OfferRepository.cs [hunk 17]
```diff
@@ -37,18 +37,18 @@ public class AppRepository : IAppRepository
...
     /// </summary>
     /// <param name="portalDbContext">PortalDb context.</param>
-    public AppRepository(PortalDbContext portalDbContext)
+    public OfferRepository(PortalDbContext portalDbContext)
     {
         this._context = portalDbContext;
     }
 
     /// <inheritdoc />
     public Task<bool> CheckAppExistsById(Guid appId) =>
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/OfferRepository.cs:46]: valid point. added the separate method for service

File: src/services/CatenaX.NetworkServices.Service.Service/BusinessLogic/IServiceBusinessLogic.cs (hunk 7 above)
Comment (Phil91) [reply src/services/CatenaX.NetworkServices.Service.Service/BusinessLogic/IServiceBusinessLogic.cs]: good idea - added

File: src/services/CatenaX.NetworkServices.Service.Service/CatenaX.NetworkServices.Service.Service.csproj (hunk 8 above)
Comment (Phil91) [reply src/services/CatenaX.NetworkServices.Service.Service/CatenaX.NetworkServices.Service.Service.csproj:16]: i've cleaned up the dependencies

File: src/services/CatenaX.NetworkServices.Service.Service/Controllers/ServiceController.cs [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+    [HttpPost]
+    [Route("{serviceId}/subscribe")]
+    [Authorize(Roles = "subscribe_service")]
+    [ProducesResponseType(typeof(NoContentResult), StatusCodes.Status204NoContent)]
+    [ProducesResponseType(typeof(ErrorResponse), StatusCodes.Status400BadRequest)]
+    [ProducesResponseType(typeof(ErrorResponse), StatusCodes.Status404NotFound)]
+    public async Task<ActionResult<Guid>> AddServiceSubscription([FromRoute] Guid serviceId)
+    {
+        var serviceSubscriptionId = await this.WithIamUserId(iamUserId => _serviceBusinessLogic.AddServiceSubscription(serviceId, iamUserId)).ConfigureAwait(false);
+        return CreatedAtRoute(nameof(GetSubscriptionDetail), new { notificationId = serviceSubscriptionId }, serviceSubscriptionId);
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/services/CatenaX.NetworkServices.Service.Service/Controllers/ServiceController.cs]: fixed

File: src/services/CatenaX.NetworkServices.Service.Service/Controllers/ServiceController.cs [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+    /// <param name="data">The data for the new service offering.</param>
+    /// <remarks>Example: POST: /api/services/addservice</remarks>
+    /// <response code="200">Returns the newly created service id.</response>
+    /// <response code="400">The given service offering data were invalid.</response>
+    [HttpPost]
+    [Route("addservice")]
+    [Authorize(Roles = "add_service_offering")]
+    [ProducesResponseType(typeof(Guid), StatusCodes.Status200OK)]
+    [ProducesResponseType(typeof(ErrorResponse), StatusCodes.Status400BadRequest)]
+    public Task<Guid> CreateServiceOffering([FromBody] ServiceOfferingData data) =>
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/services/CatenaX.NetworkServices.Service.Service/Controllers/ServiceController.cs]: fixed

**Review by ntruchsess — CHANGES_REQUESTED** [review]

**Review by ntruchsess — APPROVED** [review]

**Review by ntruchsess — CHANGES_REQUESTED** [review]
need to rereview as DBAccess-tests do fail now

**Comment by ntruchsess** [comment]
ok - I'm done so far, but in the devenv-PR https://github.com/tractusx-team-portal-onboarding/devenv/pull/114 we have 'id uuid DEFAULT gen_random_uuid() NOT NULL,' in table 'offer_subscriptions'. This auto-creation of id is not configured in other regular entities and I cannot find why it is the case here.
As one can see in the Migration Designer-files all regular entities have ValueGeneratedOnAdd() set, but in the database-export this apparently did not create the default value creation as it does in offer_subscriptions. Any idea why this is the case?

**Comment by Phil91** [comment]
> ok - I'm done so far, but in the devenv-PR [tractusx-team-portal-onboarding/devenv#114](https://github.com/tractusx-team-portal-onboarding/devenv/pull/114) we have 'id uuid DEFAULT gen_random_uuid() NOT NULL,' in table 'offer_subscriptions'. This auto-creation of id is not configured in other regular entities and I cannot find why it is the case here. As one can see in the Migration Designer-files all regular entities have ValueGeneratedOnAdd() set, but in the database-export this apparently did not create the default value creation as it does in offer_subscriptions. Any idea why this is the case?

As we found out, came with an previous migration. It's removed now.

**Review by ntruchsess — APPROVED** [review]

**Pull request merged**